src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternPart.cs

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Dispatcher.Patterns
+{
+    public abstract class RoutePatternPart
+    {
+        // This class is not an extensibility point. It is abstract so we can extend it
+        // or add semantics later inside the library.
+        internal RoutePatternPart()
+        {
+        }
+
+        public static RoutePatternLiteral CreateLiteral(string content)
+        {
+            if (string.IsNullOrEmpty(content))
+            {
+                throw new ArgumentException(Resources.Argument_NullOrEmpty, nameof(content));
+            }
+
+            if (content.IndexOf('?') >= 0)
```

Comment:
Why is this looking for '?' specifically?

Reply:
`?`  is the delimiter for a query string - you can't put that in a path segment